src/lazy/encoder/write_as_ion.rs

```diff
@@ -0,0 +1,273 @@
+//! Defines traits that allow Rust values to be serialized as Ion.
+//!
+//! In order for a [`LazyRawWriter`](crate::lazy::encoder::LazyRawWriter) to serialize a Rust value
+//! as Ion, that Rust type must implement [`WriteAsIonValue`] and may optionally also
+//! implement [`WriteAsIon`].
+//!
+//! [`WriteAsIonValue`] allows the implementor to map the Rust value to an unannotated Ion value.
+//!
+//! [`WriteAsIon`] builds on [`WriteAsIonValue`], offering a staged writer that requires the
+//! implementor to specify what annotations to write before delegating to [`WriteAsIonValue`]
+//! to serialize the value itself.
+//!
+//! Types that do not explicitly implement [`WriteAsIon`] will fall back to a blanket implementation
+//! that uses an empty annotations sequence. A custom annotations sequence can be set on a per-value
+//! basis by using the [`annotate`](crate::lazy::encoder::annotate::Annotate::annotate) method
+//! provided by the [`Annotate`](crate::lazy::encoder::annotate::Annotate) trait.
+use crate::lazy::encoder::{AnnotatedValueWriter, LazyEncoder, SequenceWriter, ValueWriter};
+use crate::{
+    Blob, Clob, Decimal, Int, IonResult, Null, RawSymbolToken, RawSymbolTokenRef, Symbol,
+    SymbolRef, Timestamp,
+};
+use std::io::Write;
+
+/// Defines how a Rust type should be serialized as Ion in terms of the methods available
+/// on [`ValueWriter`].
+pub trait WriteAsIonValue {
+    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
+        &self,
+        writer: V,
+    ) -> IonResult<()>;
+}
+
+/// Defines how a Rust type should be serialized as Ion in terms of the methods available
+/// on [`AnnotatedValueWriter`] and [`ValueWriter`].
+pub trait WriteAsIon {
+    fn write_as_ion<'a, W: Write + 'a, E: LazyEncoder<W>, V: AnnotatedValueWriter<'a, W, E>>(
+        &self,
+        writer: V,
+    ) -> IonResult<()>;
+}
+
+// Any type that does not define `WriteAsIon` itself will use this blanket implementation that does
+// not write any annotations.
+impl<T> WriteAsIon for T
+where
+    T: WriteAsIonValue,
+{
+    fn write_as_ion<'a, W: Write + 'a, E: LazyEncoder<W>, V: AnnotatedValueWriter<'a, W, E>>(
+        &self,
+        writer: V,
+    ) -> IonResult<()> {
+        self.write_as_ion_value(writer.no_annotations())
+    }
+}
+
+// ===== WriteAsIonValue implementations for common types =====
+
+impl WriteAsIonValue for Null {
+    fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
+        &self,
+        writer: V,
+    ) -> IonResult<()> {
+        writer.write_null(self.0)
+    }
+}
```

Comment:
You could DRY up a bit with a macro...
```rust
macro_rules! impl_write_as_ion_value {
    ($t:ty, $e:expr) => {
        impl WriteAsIonValue for $t {
            fn write_as_ion_value<'a, W: Write + 'a, E: LazyEncoder<W>, V: ValueWriter<'a, W, E>>(
                &self,
                writer: V,
            ) -> IonResult<()> {
                $e
            }
        }
    };
}

impl_write_as_ion_value!(bool, writer.write_bool(*self));
impl_write_as_ion_value!(i32, writer.write_i64(*self as i64));
impl_write_as_ion_value!(i64, writer.write_i64(*self));
impl_write_as_ion_value!(Decimal, writer.write_decimal(self));
impl_write_as_ion_value!(Timestamp, writer.write_timestamp(self));
// etc.
```
Thoughts?

Reply:
I'm for using a macro, but as written this won't work due to hygiene rules. When the compiler sees `writer.write_bool(...)`, it looks for a `writer` variable in the current scope.